pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/index/IndexPlugin.java

```diff
@@ -0,0 +1,25 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.pinot.segment.spi.index;
+
+
+public interface IndexPlugin<T extends IndexType<?, ?, ?>> {
```

Comment:
Can you add javadocs on what this interface is and how it should be used? I think in the PEP you talked about this. 

One thing that is not clear to me is: Does every IndexType require an implementation of this interface? If yes, what is the benefit of doing so?

Reply:
IndexPlugin is the way indexes are discovered. I preferred to separate the responsibilities in different classes, but the implementator can have both in the same class by implementing both IndexType and IndexPlugin.